Incorrect too-many-function-args for a function defined and called within a class

### Bug description

```python
class Dyn:
    def _make_method(name):     # pylint: disable=no-self-argument
        def _the_method(self):
            print(name)
        return _the_method

    print_nedbat = _make_method("nedbat")

Dyn().print_nedbat()
```
This code works. But pylint doesn't like it:
```
% pylint -E weirdmethod.py
************* Module weirdmethod
weirdmethod.py:7:19: E1121: Too many positional arguments for method call (too-many-function-args)
```

### Configuration

_No response_

### Command used

```shell
pylint -E weirdmethod.py
```


### Pylint output

```shell
************* Module weirdmethod
weirdmethod.py:7:19: E1121: Too many positional arguments for method call (too-many-function-args)
```


### Expected behavior

Since the code works, this is not an error. Pylint shouldn't complain.

### Pylint version

```shell
pylint 2.13.8
astroid 2.11.5
Python 3.7.13 (default, Mar 16 2022, 20:45:04)
[Clang 12.0.0 (clang-1200.0.32.29)]

(btw, this didn't happen with pylint 2.13.7 and astroid 2.11.3)
```


### OS / Environment

Mac OS 10.15.7

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/6592.false_positive
+++ b/doc/whatsnew/fragments/6592.false_positive
@@ -1,0 +1,3 @@
+Fix false positive for ``too-many-function-args`` when a function call is assigned to a class attribute inside the class where the function is defined.
+
+Closes #6592

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -1458,6 +1458,19 @@ accessed. Python regular expressions are accepted.",
         keyword_args += list(already_filled_keywords)
         num_positional_args += implicit_args + already_filled_positionals
 
+        # Decrement `num_positional_args` by 1 when a function call is assigned to a class attribute
+        # inside the class where the function is defined.
+        # This avoids emitting `too-many-function-args` since `num_positional_args`
+        # includes an implicit `self` argument which is not present in `called.args`.
+        if (
+            isinstance(node.frame(), nodes.ClassDef)
+            and isinstance(node.parent, (nodes.Assign, nodes.AnnAssign))
+            and isinstance(called, nodes.FunctionDef)
+            and called in node.frame().body
+            and num_positional_args > 0
+        ):
+            num_positional_args -= 1
+
         # Analyze the list of formal parameters.
         args = list(itertools.chain(called.args.posonlyargs or (), called.args.args))
         num_mandatory_parameters = len(args) - len(called.args.defaults)

--- a/tests/functional/a/arguments.py
+++ b/tests/functional/a/arguments.py
@@ -1,6 +1,6 @@
 # pylint: disable=too-few-public-methods, missing-docstring,import-error,wrong-import-position
 # pylint: disable=wrong-import-order, unnecessary-lambda, consider-using-f-string
-# pylint: disable=unnecessary-lambda-assignment
+# pylint: disable=unnecessary-lambda-assignment, no-self-argument, unused-argument
 
 def decorator(fun):
     """Decorator"""
@@ -261,3 +261,20 @@ def func(one, two, three):
 
 
 CALL = lambda *args: func(*args)
+
+
+# Ensure `too-many-function-args` is not emitted when a function call is assigned
+# to a class attribute inside the class where the function is defined.
+# Reference: https://github.com/PyCQA/pylint/issues/6592
+class FruitPicker:
+    def _pick_fruit(fruit):
+        def _print_selection(self):
+            print(f"Selected: {fruit}!")
+        return _print_selection
+
+    pick_apple = _pick_fruit("apple")
+    pick_pear = _pick_fruit("pear")
+
+picker = FruitPicker()
+picker.pick_apple()
+picker.pick_pear()

